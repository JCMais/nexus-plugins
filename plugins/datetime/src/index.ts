import {
  core,
  plugin,
  dynamicOutputMethod,
  objectType,
  stringArg,
  interfaceType,
  nonNull,
} from 'nexus'

import * as moment from 'moment-timezone'

export type DateTimePluginConfig = {
  /**
   * The name of the dateTime field, defaults to "dateTime"
   */
  dateTimeFieldName?: string

  nexusSchemaImportId?: string
  dateTimePluginImportId?: string
}

export type DateTimePluginFieldConfig<
  TypeName extends string = any,
  FieldName extends string = any
> = {
  /**
   * This defaults to the fieldName used when calling dateTime(fieldName)
   */
  dateTimeISOField?: string
} & Exclude<NexusGenPluginFieldConfig<TypeName, FieldName>, 'resolve' | 'type'>

export function dateTimePlugin(pluginConfig: DateTimePluginConfig = {}) {
  const {
    dateTimeFieldName = 'dateTime',
    dateTimePluginImportId = '@jcm/nexus-plugin-datetime',
  } = pluginConfig

  return plugin({
    name: 'DateTime',
    description: 'add t.dateTime(field) to the schema builder',
    // we want to add a extension
    fieldDefTypes: [
      core.printedGenTypingImport({
        module: dateTimePluginImportId,
        bindings: ['DateTimePluginFieldConfig'],
      }),
    ],
    onInstall(builder) {
      builder.addType(
        dynamicOutputMethod({
          name: dateTimeFieldName,
          typeDefinition: `<FieldName extends string>(
            fieldName: FieldName, 
            config: DateTimePluginFieldConfig<TypeName, FieldName>
          ): void`,
          factory({ typeName: _parentTypeName, typeDef: t, args: factoryArgs }) {
            const [fieldName, fieldConfig = {}] = factoryArgs
            const { dateTimeISOField = fieldName, ...remainingFieldConfig } = fieldConfig

            if (!builder.hasType('DateTimeFieldInterface')) {
              builder.addType(
                interfaceType({
                  name: 'DateTimeFieldInterface',
                  description: 'A object that represents an ISO datetime',
                  resolveType(obj) {
                    if (obj.iso) return 'DateTimeField'
                    return null
                  },
                  definition(t2) {
                    t2.nonNull.string('iso')
                    t2.nonNull.string('formatted', {
                      args: {
                        format: nonNull(stringArg()),
                        timezone: stringArg({
                          description: 'Timezone to format the ISO date, defaults to UTC',
                        }),
                      },
                      resolve: (root, args) => {
                        return args.timezone
                          ? moment.tz(root.iso, args.timzone).format(args.format)
                          : moment.utc(root.iso).format(args.format)
                      },
                    })
                    t2.nonNull.boolean('isAfter', {
                      args: {
                        iso: stringArg({
                          description: 'Defaults to the current time if not provided',
                        }),
                      },
                      resolve: (root, { iso }) => moment(root.iso).isAfter(iso || new Date()),
                    })
                    t2.nonNull.boolean('isBefore', {
                      args: {
                        iso: stringArg({
                          description: 'Defaults to the current time if not provided',
                        }),
                      },
                      resolve: (root, { iso }) => moment(root.iso).isBefore(iso || new Date()),
                    })
                    t2.nonNull.boolean('isSameOrAfter', {
                      args: {
                        iso: stringArg({
                          description: 'Defaults to the current time if not provided',
                        }),
                      },
                      resolve: (root, { iso }) => moment(root.iso).isSameOrAfter(iso || new Date()),
                    })
                    t2.nonNull.boolean('isSameOrBefore', {
                      args: {
                        iso: stringArg({
                          description: 'Defaults to the current time if not provided',
                        }),
                      },
                      resolve: (root, { iso }) =>
                        moment(root.iso).isSameOrBefore(iso || new Date()),
                    })
                    t2.nonNull.boolean('isSame', {
                      args: {
                        iso: stringArg({
                          description: 'Defaults to the current time if not provided',
                        }),
                      },
                      resolve: (root, { iso }) => moment(root.iso).isSame(iso || new Date()),
                    })
                    t2.nonNull.boolean('isBetween', {
                      args: {
                        isoStart: stringArg({
                          description: 'Defaults to the start of the current day if not provided',
                        }),
                        isoEnd: stringArg({
                          description: 'Defaults to the end of the current day if not provided',
                        }),
                      },
                      resolve: (root, { isoStart, isoEnd }) =>
                        moment(root.iso).isBetween(
                          isoStart || moment().startOf('day'),
                          isoEnd || moment().endOf('day'),
                        ),
                    })
                  },
                }),
              )
            }

            if (!builder.hasType('DateTimeField')) {
              builder.addType(
                objectType({
                  name: 'DateTimeField',
                  description: 'Represents an ISO datetime that can be formatted in other formats',
                  definition: (t2) => {
                    t2.implements('DateTimeFieldInterface')
                  },
                }),
              )
            }

            t.field(fieldName, {
              ...remainingFieldConfig,
              type: 'DateTimeField',
              resolve: (root) =>
                root[dateTimeISOField] && {
                  iso: moment(root[dateTimeISOField]).toISOString(),
                },
            })
          },
        }),
      )
    },
  })
}
