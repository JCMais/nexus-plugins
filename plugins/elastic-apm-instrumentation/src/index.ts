import { plugin, core } from '@nexus/schema'
import {
  isLeafType,
  isWrappingType,
  ResponsePath,
  GraphQLOutputType,
  GraphQLWrappingType,
} from 'graphql'
import type * as Agent from 'elastic-apm-node'

// Most of this code is inspired on the authorize plugin from @nexus/schema

const fieldDefTypes = core.printedGenTyping({
  optional: true,
  name: 'shouldCreateApmSpan',
  description: `
    This can be used to override the elastic apm instrumentation settings and force
    a span to [not] be created for this field.
  `,
  type: 'boolean',
})

export interface ElasticApmInstrumentationPluginConfig {
  apmAgent: typeof Agent

  /**
   * Should create spans by default for all fields that pass the ignore checks?
   *
   * Defaults to true
   */
  shouldCreateApmSpanByDefault?: boolean

  /**
   * Should leaf fields be ignored?
   *
   * Defaults to true
   */
  shouldIgnoreLeafFieldsByDefault?: boolean

  /**
   * Pass an array of fields to be ignored from span creation, each item can be a string or a RegExp
   * If a string, it's also possible to use wildcards, like: `PageInfo.*`
   */
  ignoredFields?: Array<string | RegExp>

  /**
   * Pass an array of types to be ignored from span creation, each item can be a string or a RegExp
   */
  ignoredTypes?: Array<string | RegExp>
}

const isValidMatch = (strA: string, strB: string) => strA === strB || strA === '*'

const isPatternMatchingField = (pattern: string, field: string) => {
  if (pattern === '*') return true

  const piecesPatternPieces = pattern.split('.')
  const fieldPatternPieces = field.split('.')

  if (piecesPatternPieces.length !== 2 || fieldPatternPieces.length !== 2) {
    throw new Error('Pattern/Field path must have exactly 2 parts, ParentType.childType')
  }

  return (
    isValidMatch(piecesPatternPieces[0], fieldPatternPieces[0]) &&
    isValidMatch(piecesPatternPieces[1], fieldPatternPieces[1])
  )
}

const getFirstNonWrappingType = (
  parent: GraphQLOutputType,
): Exclude<GraphQLOutputType, GraphQLWrappingType> => {
  if (isWrappingType(parent)) return getFirstNonWrappingType(parent.ofType)

  return parent
}

const getGraphQLFieldPath = (path?: ResponsePath): string =>
  (path
    ? [
        getGraphQLFieldPath(path.prev),
        // @ts-expect-error
        path.prev ? path.key : `${path.typename}.${path.key}`,
      ].filter((v) => !!v)
    : []
  ).join('.')

export const elasticApmInstrumentationPlugin = (
  pluginConfig: ElasticApmInstrumentationPluginConfig,
) => {
  const {
    apmAgent,
    shouldCreateApmSpanByDefault = true,
    shouldIgnoreLeafFieldsByDefault = true,
    ignoredFields = [],
    ignoredTypes = [],
  } = pluginConfig

  return plugin({
    name: 'ElasticApmReporter',
    description: 'Creates separated spans for each resolved field',
    fieldDefTypes: fieldDefTypes,
    onCreateFieldResolver(config) {
      const firstNonWrappingType = getFirstNonWrappingType(config.fieldConfig.type)
      const isLeaf = isLeafType(firstNonWrappingType)

      const parentTypeName = config.parentTypeConfig.name
      const fieldName = config.fieldConfig.name
      const qualifiedName = `${parentTypeName}.${fieldName}`

      const isIgnoredField = ignoredFields.some((ignoredField) =>
        ignoredField instanceof RegExp
          ? ignoredField.test(qualifiedName)
          : isPatternMatchingField(ignoredField, qualifiedName),
      )

      const isIgnoredType = ignoredTypes.some((type) =>
        type instanceof RegExp
          ? type.test(firstNonWrappingType.name)
          : type === firstNonWrappingType.name,
      )

      const shouldCreateApmSpan =
        config.fieldConfig.extensions?.nexus?.config?.shouldCreateApmSpan ??
        (shouldCreateApmSpanByDefault &&
          (!shouldIgnoreLeafFieldsByDefault || !isLeaf) &&
          !isIgnoredField &&
          !isIgnoredType)

      // console.log({
      //   qualifiedName,
      //   shouldCreateApmSpan,
      //   firstNonWrappingType,
      //   isLeaf,
      //   isIgnoredField,
      //   isIgnoredType,
      //   type: config.fieldConfig.type,
      // })

      if (!shouldCreateApmSpan) {
        return
      }

      return function (root, args, ctx, info, next) {
        const isApmAgentStartedAndHasTransaction =
          apmAgent.isStarted() && apmAgent.currentTransaction

        if (!isApmAgentStartedAndHasTransaction) {
          return next(root, args, ctx, info)
        }

        const fieldPath = getGraphQLFieldPath(info.path)
        const span = apmAgent.startSpan(
          `GraphQL resolver: ${fieldPath}`,
          'db',
          'graphql',
          'resolve',
        )

        let toComplete
        try {
          toComplete = next(root, args, ctx, info)
        } catch (e) {
          toComplete = Promise.reject(e)
        }
        return plugin.completeValue(
          toComplete,
          (val) => {
            span?.end()
            return val
          },
          (error) => {
            span?.end()
            throw error
          },
        )
      }
    },
  })
}
