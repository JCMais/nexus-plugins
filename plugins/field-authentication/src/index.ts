import { plugin, core } from '@nexus/schema'
import { GraphQLResolveInfo } from 'graphql'
import { MaybePromise } from '@nexus/schema/dist/core'

// Most of this code is inspired on the authorize plugin from @nexus/schema

const fieldAuthenticationPluginResolverImport = core.printedGenTypingImport({
  module: '@jcm/nexus-plugin-field-authentication',
  bindings: ['FieldAuthenticationResolver'],
})

const fieldDefTypes = core.printedGenTyping({
  optional: true,
  name: 'authentication',
  description: `
    Authentication for an individual field. Returning "true"
    or "Promise<true>" means the field can be accessed only if
    the request is authenticated. Returning "false" or "Promise<false>"
    means the field can be acessed only if the request is *NOT* authenticated.
    It's also possible to return a tuple "[boolean, Error | ResolveValue]"
    where the first element specifies if the request should be authenticated
    or not and the second specifies the behavior:
    If the second element is an error, this error will be returned for the field, even if
      throwErrorOnFailedAuthentication is false.
    If the second element is anything else, this value will be resolved if the request
     fails the authentication check.
  `,
  type: 'FieldAuthenticationResolver<TypeName, FieldName>',
  imports: [fieldAuthenticationPluginResolverImport],
})

export type FieldAuthenticationResolverReturnValue =
  | boolean
  | Error
  | [boolean, Error]
  | [boolean, core.ResultValue<any, any>]

export type FieldAuthenticationResolver<TypeName extends string, FieldName extends string> = (
  root: core.RootValue<TypeName>,
  args: core.ArgsValue<TypeName, FieldName>,
  context: core.GetGen<'context'>,
  info: GraphQLResolveInfo,
) => core.MaybePromise<FieldAuthenticationResolverReturnValue>

export interface FieldAuthenticationPluginErrorConfig {
  error: Error
  root: any
  args: any
  ctx: core.GetGen<'context'>
  info: GraphQLResolveInfo
}

export const defaultFormatError = ({ error }: FieldAuthenticationPluginErrorConfig): Error => error

export type FieldAuthenticationPluginConfigIsLogged<
  TypeName extends string,
  FieldName extends string
> = (
  root: core.RootValue<TypeName>,
  args: core.ArgsValue<TypeName, FieldName>,
  context: core.GetGen<'context'>,
  info: GraphQLResolveInfo,
) => boolean | MaybePromise<boolean>

export interface FieldAuthenticationPluginConfig {
  /**
   * This will be called with the error that will be returned when resolving a field
   *  that did not pass the authentication check when throwErrorOnFailedAuthenticationByDefault
   *  was set to true or when the authentication had a tuple [boolean, Error]
   * By default this is the function ({error: Error}) => error
   */
  formatError?: (authConfig: FieldAuthenticationPluginErrorConfig) => Error
  /**
   * If this is true and the authentication field does not include
   *  a default resolve value, an error will be thrown on failed authentication
   */
  throwErrorOnFailedAuthenticationByDefault?: boolean
  /**
   * If throwErrorOnFailedAuthenticationByDefault is true,
   */
  defaultErrorMessage?: string
  /**
   * If throwErrorOnFailedAuthenticationByDefault is false, a failed authentication check will return
   *  this value as the resolved value of the field
   */
  defaultResolveValue?: core.ResultValue<any, any>
  /**
   * By default this checks if ctx.state.user exists
   */
  isLogged?: FieldAuthenticationPluginConfigIsLogged<any, any>
}

// This is much more complex than it could be
//  but it was done that way so that it handle all scenarios possible that we might ever encounter.

export function fieldAuthenticationPlugin(pluginConfig: FieldAuthenticationPluginConfig = {}) {
  const {
    formatError = defaultFormatError,
    throwErrorOnFailedAuthenticationByDefault = false,
    defaultErrorMessage = 'Not Authenticated',
    defaultResolveValue = null,
    isLogged = (_root, _args, ctx, _info) => !!ctx?.state?.user,
  } = pluginConfig

  const ensureError = (
    root: any,
    args: any,
    ctx: core.GetGen<'context'>,
    info: GraphQLResolveInfo,
  ) => (error: Error) => {
    const finalErr = formatError({ error, root, args, ctx, info })
    if (finalErr instanceof Error) {
      throw finalErr
    }
    ;(ctx.logger || console).error(
      `Non-Error value ${JSON.stringify(
        finalErr,
      )} returned from custom formatError in field authentication plugin`,
    )
    throw new Error(defaultErrorMessage)
  }

  return plugin({
    name: 'FieldAuthentication',
    // we want to add a extension
    description: 'Makes sure request is authenticated before calling next resolvers in the chain',
    fieldDefTypes: fieldDefTypes,
    onCreateFieldResolver(config) {
      const authentication = config.fieldConfig.extensions?.nexus?.config?.authentication

      if (typeof authentication !== 'undefined') {
        // The authentication wrapping resolver.
        return function (root, args, ctx, info, next) {
          const {
            fieldName,
            parentType: { name: parentTypeName },
          } = info

          const processAuthenticationResult = (
            isUserLogged: boolean,
            result: FieldAuthenticationResolverReturnValue,
          ) => {
            const finalFormatError = ensureError(root, args, ctx, info)

            if (
              typeof result === 'boolean' ||
              (Array.isArray(result) && result.length === 2 && typeof result[0] === 'boolean')
            ) {
              const canProceed =
                isUserLogged === result || (Array.isArray(result) && isUserLogged === result[0])

              // all branches here must return finalFormatError
              if (!canProceed) {
                if (Array.isArray(result)) {
                  if (result[1] instanceof Error) {
                    return finalFormatError(result[1])
                  }
                  return result[1]
                } else if (throwErrorOnFailedAuthenticationByDefault) {
                  return finalFormatError(new Error(defaultErrorMessage))
                }
                return defaultResolveValue
              }

              return next(root, args, ctx, info)
            } else {
              return finalFormatError(
                new Error(
                  `Field authentication for ${parentTypeName}.${fieldName} expected a boolean or [boolean, resolvedValue] tuple, saw ${JSON.stringify(
                    result,
                  )} instead`,
                ),
              )
            }
          }

          let toCompleteIsLogged
          try {
            toCompleteIsLogged = isLogged(root, args, ctx, info)
          } catch (e) {
            toCompleteIsLogged = Promise.reject(e)
          }
          return plugin.completeValue(
            toCompleteIsLogged,
            (isUserLogged) => {
              if (typeof authentication !== 'function') {
                return processAuthenticationResult(isUserLogged, authentication)
              }

              let toComplete
              try {
                toComplete = authentication(root, args, ctx, info)
              } catch (e) {
                toComplete = Promise.reject(e)
              }
              return plugin.completeValue(
                toComplete,
                processAuthenticationResult.bind(undefined, isUserLogged),
                // @ts-expect-error
                (err) => {
                  ensureError(root, args, ctx, info)(err as Error)
                },
              )
            },
            // @ts-expect-error
            (err) => {
              ensureError(root, args, ctx, info)(err)
            },
          )
        }
      }
    },
  })
}
