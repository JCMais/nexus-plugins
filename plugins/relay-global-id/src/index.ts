import { GraphQLResolveInfo } from 'graphql'
import { plugin, dynamicOutputMethod, core } from 'nexus'

import { toGlobalId } from 'graphql-relay'

type RelayGlobalIdConfigResolveFunction<TypeName extends string, FieldName extends string> = (
  root: core.RootValueField<TypeName, FieldName>,
  args: core.ArgsValue<TypeName, FieldName>,
  ctx: core.GetGen<'context'>,
  info: GraphQLResolveInfo,
) =>
  | core.MaybePromise<core.ResultValue<TypeName, FieldName>>
  | core.MaybePromiseDeep<core.ResultValue<TypeName, FieldName>>

export type RelayGlobalIdPluginConfig<
  TypeName extends string = any,
  FieldName extends string = any
> = {
  nexusFieldName?: string
  nexusSchemaImportId?: string
  relayGlobalIdPluginImportId?: string

  /**
   * Defaults to true - Adds a new ID! field called `raw${uppercase(fieldName)}`, where `fieldName` is `t.relayGlobalId(fieldName)`,
   *  with the original value of the field.
   *
   * It's also possible to pass a string with a different name for the field
   *
   * You can also set this in a per field basis
   */
  shouldAddRawId?: boolean | string

  /**
   * If no resolve is supplied, the default value to be used as global ID will be root[field]
   *
   * This defaults to the fieldName used when calling relayGlobalId(fieldName)
   *
   * If a resolve is passed, this is ignored
   *
   * You can also set this in a per field basis
   */
  field?: string

  /**
   * You can use this to specificy your own resolve function for the ID
   *
   * You can also set this in a per field basis
   */
  resolve?: RelayGlobalIdConfigResolveFunction<TypeName, FieldName>
}

export type RelayGlobalIdNexusFieldConfig<
  TypeName extends string = any,
  FieldName extends string = any
> = {
  /**
   * If no resolve is supplied, the default value to be used as global ID will be root[field]
   *
   * This defaults to the fieldName used when calling relayGlobalId(fieldName)
   *
   * If a resolve is passed, this is ignored
   */
  field?: string

  /**
   * Defaults to true - Adds a new ID! field called `raw${uppercase(fieldName)}`, where `fieldName` is `t.relayGlobalId(fieldName)`,
   *  with the original value of the field.
   *
   * It's also possible to pass a string with a different name for the field
   */
  shouldAddRawId?: boolean | string

  /**
   * Defaults to the parent type of the current field.
   */
  typeName?: string

  /**
   * You can use this to specificy your own resolve function for the ID
   */
  resolve?: RelayGlobalIdConfigResolveFunction<TypeName, FieldName>
} & NexusGenPluginFieldConfig<TypeName, FieldName>

export function relayGlobalIdPlugin(pluginConfig: RelayGlobalIdPluginConfig = {}) {
  const {
    nexusFieldName = 'relayGlobalId',
    nexusSchemaImportId = 'nexus',
    relayGlobalIdPluginImportId = '@jcm/nexus-plugin-relay-global-id',
    shouldAddRawId: shouldAddRawIdPluginConfig = true,
    field: fieldPluginConfig,
    resolve: resolvePluginConfig,
  } = pluginConfig

  return plugin({
    name: 'RelayGlobalId',
    description: 'add t.relayGlobalId(field) to the schema builder',
    fieldDefTypes: [
      core.printedGenTypingImport({
        module: nexusSchemaImportId,
        bindings: ['core'],
      }),
      core.printedGenTypingImport({
        module: relayGlobalIdPluginImportId,
        bindings: ['RelayGlobalIdNexusFieldConfig'],
      }),
    ],
    // we want to add a extension
    onInstall(builder) {
      builder.addType(
        dynamicOutputMethod({
          name: nexusFieldName,
          typeDefinition: `<FieldName extends string>(
            fieldName: FieldName, 
            config: RelayGlobalIdNexusFieldConfig<TypeName, FieldName>
          ): void`,
          factory({ typeName: parentTypeName, typeDef: t, args: factoryArgs }) {
            const [fieldName, fieldConfig = {}] = factoryArgs as [
              string,
              RelayGlobalIdNexusFieldConfig,
            ]

            const {
              field = fieldPluginConfig || fieldName,
              shouldAddRawId = shouldAddRawIdPluginConfig,
              typeName = parentTypeName,
              resolve: resolveFn = resolvePluginConfig,
              ...remainingFieldConfig
            } = fieldConfig

            t.id(fieldName, {
              ...remainingFieldConfig,
              async resolve(root, args, ctx, info) {
                const resolved = resolveFn ? await resolveFn(root, args, ctx, info) : root[field]
                return resolved && toGlobalId(typeName, resolved)
              },
            })

            if (shouldAddRawId) {
              t.id(
                typeof shouldAddRawId === 'string'
                  ? shouldAddRawId
                  : `raw${fieldName.charAt(0).toUpperCase()}${fieldName.slice(1)}`,
                {
                  resolve: resolveFn ? resolveFn : (root) => root[field],
                },
              )
            }
          },
        }),
      )
    },
  })
}
