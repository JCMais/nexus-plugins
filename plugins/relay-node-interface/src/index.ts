import { plugin, core, interfaceType, queryField, idArg } from '@nexus/schema'
import { fromGlobalId } from 'graphql-relay'
import { GraphQLResolveInfo } from 'graphql'

export type RelayNodeInterfacePluginConfig = {
  idFetcher: (
    val: { id: string; type: core.AllOutputTypes },
    ctx: core.GetGen<'context'>,
    info: GraphQLResolveInfo,
  ) => core.ResultValue<any, any>
  resolveType: (object: core.ResultValue<any, any>) => core.AllOutputTypes
  /**
   * Used to parse the ID before calling idFetcher - By default this calls fromGlobalId from graphql-relay
   */
  idParser?: (id: string) => any
}

// Pretty much based on nodeDefinitions function from
//  relay: https://github.com/graphql/graphql-relay-js/blob/8f4ed1ad35805ef233ad9fd1af33abb9c0cad35a/src/node/node.js
export function relayNodeInterfacePlugin(pluginConfig: RelayNodeInterfacePluginConfig) {
  const { idFetcher, resolveType, idParser = fromGlobalId } = pluginConfig

  if (!idFetcher) {
    throw new Error('idFetcher option is required for relayNodeInterfacePlugin')
  }

  if (typeof idFetcher !== 'function') {
    throw new Error(
      'idFetcher option must be a function with signature: async ({ id, type }, ctx, info) => T where T is the resolved value for the ID',
    )
  }

  if (!resolveType) {
    throw new Error('resolveType option is required for relayNodeInterfacePlugin')
  }

  if (typeof resolveType !== 'function') {
    throw new Error(
      'resolveType option must be an function with signature async (value) => T where T is the type name or object',
    )
  }

  return plugin({
    name: 'RelayNodeInterface',
    description: 'Creates the Relay Node interface and add node/nodes fields to the Query type',
    onInstall(builder) {
      if (!builder.hasType('Node')) {
        // node interface
        builder.addType(
          interfaceType({
            name: 'Node',
            description: 'An object with a global ID',
            definition: (t) => {
              t.id('id', {
                description: 'The global ID of the object.',
              })

              // overwrite the resolve type with the client function
              // resolveType is just one way to map objs -> to the graphql type they represent
              //  we could also have simply relied on isTypeOf on the GraphQLObjectType themselves
              //  but as relay uses this by default, let's keep using it
              t.resolveType(resolveType)
            },
          }),
        )

        // node field
        builder.addType(
          queryField((t) => {
            t.field('node', {
              type: 'Node',
              nullable: true,
              args: {
                id: idArg({
                  required: true,
                  description: 'The global ID of an object',
                }),
              },
              description: 'Fetches an object given its global ID',
              resolve: (_obj, { id }, context, info) => idFetcher(fromGlobalId(id), context, info),
            })
          }),
        )

        // nodes field
        builder.addType(
          queryField((t) => {
            t.field('nodes', {
              type: 'Node',
              list: [false],
              args: {
                ids: idArg({
                  list: [true],
                  required: true,
                  description: 'The global IDs of objects',
                }),
              },
              description: 'Fetches objects given their global IDs',
              resolve: (_obj, { ids }, context, info) =>
                Promise.all(
                  // @ts-expect-error
                  ids.map((id) => Promise.resolve(idFetcher(idParser(id), context, info))),
                ),
            })
          }),
        )
      }

      // TODO: Deprecate this syntax
      return { types: [] }
    },
  })
}
