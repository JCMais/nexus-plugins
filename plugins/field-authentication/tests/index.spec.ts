import { execute, parse } from 'graphql'
import { makeSchema, objectType } from 'nexus'

import { fieldAuthenticationPlugin, FieldAuthenticationPluginConfig } from '../src'

const user = {
  id: '8a291df8-bbcf-11ea-9db0-ff6b23713451',
  name: 'user-a',
}

const User = objectType({
  name: 'User',
  definition(t) {
    t.id('id')
    t.string('name')
  },
})

const QueryAll = parse(
  `

  fragment userInfo on User {
    id
    name
  }

  query QueryAll {
    me1 {
      ...userInfo
    }
    me2 {
      ...userInfo
    }
    me3 {
      ...userInfo
    }
    me4 {
      ...userInfo
    }
    me5 {
      ...userInfo
    }
    me6 {
      ...userInfo
    }
    me7 {
      ...userInfo
    }
    me8 {
      ...userInfo
    }
    me9 {
      ...userInfo
    }
    me10 {
      ...userInfo
    }
  }`,
)

const loggedOutUser = {
  id: 'unauthenticated',
  name: 'guest',
}

const fakeUser = {
  id: 'totally-different-than-the-logged',
  name: 'another-user',
}

const testSchema = (pluginConfig: FieldAuthenticationPluginConfig, outputs = false) =>
  makeSchema({
    outputs,
    types: [
      User,
      objectType({
        name: 'Query',
        definition(t) {
          // no authentication check
          t.field('me1', {
            type: User,
            resolve: () => user,
          })
          // true authentication check
          t.field('me2', {
            type: User,
            // @ts-expect-error
            authentication: true,
            resolve: () => user,
          })
          // false authentication check
          t.field('me3', {
            type: User,
            // @ts-expect-error
            authentication: false,
            resolve: () => user,
          })
          // true authentication check - different return value
          t.field('me4', {
            type: User,
            // @ts-expect-error
            authentication: [true, loggedOutUser],
            resolve: () => user,
          })
          // false authentication check - different return value
          t.field('me5', {
            type: User,
            // @ts-expect-error
            authentication: [false, fakeUser],
            resolve: () => user,
          })
          // true authentication check - error
          t.field('me6', {
            type: User,
            // @ts-expect-error
            authentication: [true, new Error('Invalid - Unauthenticated')],
            resolve: () => user,
          })
          // false authentication check - error
          t.field('me7', {
            type: User,
            // @ts-expect-error
            authentication: [false, new Error('Invalid - Authenticated')],
            resolve: () => user,
          })
          // authentication function
          t.field('me8', {
            type: User,
            // @ts-expect-error
            authentication: () => true,
          })
          // authentication function - thrown error
          t.field('me9', {
            type: User,
            // @ts-expect-error
            authentication: () => {
              throw new Error('You shall not pass')
            },
          })
          // invalid value
          t.field('me10', {
            type: User,
            // @ts-expect-error
            authentication: { something: true },
          })
        },
      }),
    ],
    plugins: [fieldAuthenticationPlugin(pluginConfig)],
    nonNullDefaults: {
      input: false,
      output: false,
    },
  })

describe('fieldAuthentication', () => {
  describe('basic behavior', () => {
    it('should resolve all fields correctly with default plugin values - unauthenticated', async () => {
      const schema = testSchema({})
      const data = await execute({
        schema,
        document: QueryAll,
      })

      expect(data).toMatchSnapshot()
    })

    it('should resolve all fields correctly with default plugin values - authenticated', async () => {
      const schema = testSchema({})
      const data = await execute({
        schema,
        contextValue: {
          state: {
            user: 'something',
          },
        },
        document: QueryAll,
      })

      expect(data).toMatchSnapshot()
    })

    it('throwErrorOnFailedAuthenticationByDefault = true / defaultErrorMessage = something', async () => {
      const schema = testSchema({
        throwErrorOnFailedAuthenticationByDefault: true,
        defaultErrorMessage: 'You need to be authenticated',
      })
      const data = await execute({
        schema,
        contextValue: {
          state: {
            user: 'something',
          },
        },
        document: QueryAll,
      })

      expect(data).toMatchSnapshot()
    })

    it('format error returning non error value', async () => {
      const schema = testSchema({
        throwErrorOnFailedAuthenticationByDefault: true,
        defaultErrorMessage: 'You need to be authenticated',
        // @ts-expect-error
        formatError: () => "I'm not an error",
      })
      const errorLogger = jest.fn()
      const data = await execute({
        schema,
        contextValue: {
          state: {
            user: 'something',
          },
          logger: {
            error: errorLogger,
          },
        },
        document: QueryAll,
      })

      expect(data).toMatchSnapshot()
      expect(errorLogger).toHaveBeenCalled()
    })
  })
})
