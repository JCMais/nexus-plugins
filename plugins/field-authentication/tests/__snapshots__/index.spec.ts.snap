// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fieldAuthentication basic behavior format error returning non error value 1`] = `
Object {
  "data": Object {
    "me1": Object {
      "id": "8a291df8-bbcf-11ea-9db0-ff6b23713451",
      "name": "user-a",
    },
    "me10": null,
    "me2": Object {
      "id": "8a291df8-bbcf-11ea-9db0-ff6b23713451",
      "name": "user-a",
    },
    "me3": null,
    "me4": Object {
      "id": "8a291df8-bbcf-11ea-9db0-ff6b23713451",
      "name": "user-a",
    },
    "me5": Object {
      "id": "totally-different-than-the-logged",
      "name": "another-user",
    },
    "me6": Object {
      "id": "8a291df8-bbcf-11ea-9db0-ff6b23713451",
      "name": "user-a",
    },
    "me7": null,
    "me8": null,
    "me9": null,
  },
  "errors": Array [
    [GraphQLError: You need to be authenticated],
    [GraphQLError: You need to be authenticated],
    [GraphQLError: You need to be authenticated],
    [GraphQLError: You need to be authenticated],
  ],
}
`;

exports[`fieldAuthentication basic behavior should resolve all fields correctly with default plugin values - authenticated 1`] = `
Object {
  "data": Object {
    "me1": Object {
      "id": "8a291df8-bbcf-11ea-9db0-ff6b23713451",
      "name": "user-a",
    },
    "me10": null,
    "me2": Object {
      "id": "8a291df8-bbcf-11ea-9db0-ff6b23713451",
      "name": "user-a",
    },
    "me3": null,
    "me4": Object {
      "id": "8a291df8-bbcf-11ea-9db0-ff6b23713451",
      "name": "user-a",
    },
    "me5": Object {
      "id": "totally-different-than-the-logged",
      "name": "another-user",
    },
    "me6": Object {
      "id": "8a291df8-bbcf-11ea-9db0-ff6b23713451",
      "name": "user-a",
    },
    "me7": null,
    "me8": null,
    "me9": null,
  },
  "errors": Array [
    [GraphQLError: Invalid - Authenticated],
    [GraphQLError: Field authentication for Query.me10 expected a boolean or [boolean, resolvedValue] tuple, saw {"something":true} instead],
    [GraphQLError: You shall not pass],
  ],
}
`;

exports[`fieldAuthentication basic behavior should resolve all fields correctly with default plugin values - unauthenticated 1`] = `
Object {
  "data": Object {
    "me1": Object {
      "id": "8a291df8-bbcf-11ea-9db0-ff6b23713451",
      "name": "user-a",
    },
    "me10": null,
    "me2": null,
    "me3": Object {
      "id": "8a291df8-bbcf-11ea-9db0-ff6b23713451",
      "name": "user-a",
    },
    "me4": Object {
      "id": "unauthenticated",
      "name": "guest",
    },
    "me5": Object {
      "id": "8a291df8-bbcf-11ea-9db0-ff6b23713451",
      "name": "user-a",
    },
    "me6": null,
    "me7": Object {
      "id": "8a291df8-bbcf-11ea-9db0-ff6b23713451",
      "name": "user-a",
    },
    "me8": null,
    "me9": null,
  },
  "errors": Array [
    [GraphQLError: Invalid - Unauthenticated],
    [GraphQLError: Field authentication for Query.me10 expected a boolean or [boolean, resolvedValue] tuple, saw {"something":true} instead],
    [GraphQLError: You shall not pass],
  ],
}
`;

exports[`fieldAuthentication basic behavior throwErrorOnFailedAuthenticationByDefault = true / defaultErrorMessage = something 1`] = `
Object {
  "data": Object {
    "me1": Object {
      "id": "8a291df8-bbcf-11ea-9db0-ff6b23713451",
      "name": "user-a",
    },
    "me10": null,
    "me2": Object {
      "id": "8a291df8-bbcf-11ea-9db0-ff6b23713451",
      "name": "user-a",
    },
    "me3": null,
    "me4": Object {
      "id": "8a291df8-bbcf-11ea-9db0-ff6b23713451",
      "name": "user-a",
    },
    "me5": Object {
      "id": "totally-different-than-the-logged",
      "name": "another-user",
    },
    "me6": Object {
      "id": "8a291df8-bbcf-11ea-9db0-ff6b23713451",
      "name": "user-a",
    },
    "me7": null,
    "me8": null,
    "me9": null,
  },
  "errors": Array [
    [GraphQLError: You need to be authenticated],
    [GraphQLError: Invalid - Authenticated],
    [GraphQLError: Field authentication for Query.me10 expected a boolean or [boolean, resolvedValue] tuple, saw {"something":true} instead],
    [GraphQLError: You shall not pass],
  ],
}
`;
